### PR TITLE
Pager area

### PR DESCRIPTION
**Pull request by jenswi-linaro**
This is a step preparing to merge struct tee_pager_area with struct vm_region.

### CODE REVIEW

**Comment by jenswi-linaro** [comment]
I see that IBART is offline so I've just tested OK here with paging enabled on Hikey.

**Comment by jforissier** [comment]
`Tested-by: Jerome Forissier <jerome@forissier.org> (HiKey960)`

I found that commit 5ca851ec83ba ("core: pager: add struct tblidx") is broken (board hangs on boot, same on QEMUv8) and commit "core: pager: initialize vabase in core pager tables" in this PR fixes it, so perhaps a `Fixes:` tag could be added.

**Comment by jenswi-linaro** [comment]
Perhaps we should take "core: pager: initialize vabase in core pager tables" separately then since the other patch might take a while to review.

**Comment by jenswi-linaro** [comment]
I did a separate PR so we can take our time here, https://github.com/OP-TEE/optee_os/pull/4498

**Comment by jenswi-linaro** [comment]
Rebased and tested-by tag applied now that "core: pager: initialize vabase in core pager tables" was merged.

**Comment by jenswi-linaro** [comment]
Ping?

File: core/arch/arm/mm/tee_pager.c [hunk 1]
```diff
@@ -1866,8 +1940,13 @@ void tee_pager_pgt_save_and_release_entries(struct pgt *pgt)
 	areas = to_user_mode_ctx(pgt->ctx)->areas;
 	if (areas) {
 		TAILQ_FOREACH(area, areas, link) {
-			if (area->pgt == pgt)
-				area->pgt = NULL;
+			for (n = 0; n < get_pgt_count(area->base, area->size);
+			     n++) {
+				if (area->pgt_array[n] == pgt) {
+					area->pgt_array[n] = NULL;
+					break;
```
[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply core/arch/arm/mm/tee_pager.c:1947]: Each entry is supposed to be a unique pointer. See for instance `area_va2tblidx()` above.

File: core/arch/arm/mm/tee_pager.c [hunk 2]
```diff
@@ -571,57 +595,65 @@ static void area_insert(struct tee_pager_area_head *head,
...
+	if ((base & SMALL_PAGE_MASK) || !size) {
+		EMSG("invalid pager area [%" PRIxVA " +0x%zx]", base, size);
+		panic();
+	}
+
+	area = calloc(1, sizeof(*area));
+	if (!area)
+		return NULL;
+	area->pgt_array = calloc(sizeof(struct pgt *),
+				 get_pgt_count(base, size));
```
[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply core/arch/arm/mm/tee_pager.c:611]: I'll fix.

**Comment by jenswi-linaro** [comment]
Rebased on master to make sure that the tests will pass.

**Review by etienne-lms — COMMENTED** [review]
`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>`
`Tested-by: Etienne Carriere <etienne.carriere@linaro.org> (stm32mp1)`

**Comment by jenswi-linaro** [comment]
Squashed and tags applied.

**Pull request merged**